pulser/simulation/simulation.py

```diff
@@ -314,60 +369,265 @@ def get_hamiltonian(self, time: float) -> qutip.Qobj:
             extracted from the effective sequence (determined by
             `self.sampling_rate`) at the specified time.
         """
+        # Refresh the hamiltonian
+        self._build_hamiltonian_from_seq()
         if time > 1000 * self._times[-1]:
             raise ValueError("Provided time is larger than sequence duration.")
         if time < 0:
             raise ValueError("Provided time is negative.")
         return self._hamiltonian(time/1000)  # Creates new Qutip.Qobj
 
+    def _prepare_spam_detune(self) -> None:
+        """Choose atoms that will be badly prepared"""
+        # Tag True if atom is badly prepared
+        self.spam_detune: dict[Union[str, int], bool] = \
+            {qid: (np.random.uniform() < self._spam_dict['eta'])
+             for qid in self._qid_index}
+
+    @property
+    def doppler_sigma(self):
+        return self._doppler_sigma
+
+    @doppler_sigma.setter
+    def doppler_sigma(self, value: float):
+        self._doppler_sigma = value
+
+    @property
+    def temperature(self):
+        return self._temperature
+
+    @temperature.setter
+    def temperature(self, value: float):
+        # value set in microkelvin
+        self._temperature = value * 10**-6
+        self._calc_sigma_doppler()
+
+    @property
+    def spam_dict(self):
+        return self._spam_dict
```

Comment:
I would say the `property`s have a specific return type, while the `setter`s should return `None`. You tried this and it didn't work?